nexus/server.go

```diff
@@ -0,0 +1,464 @@
+package nexus
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"io"
+	"log/slog"
+	"net/http"
+	"path"
+	"time"
+
+	"github.com/gorilla/mux"
+)
+
+// StartOperationRequest is input for Handler.StartOperation.
+type StartOperationRequest struct {
+	// Operation name.
+	Operation string
+	// Request ID, should be used to dedupe start requests.
+	RequestID string
+	// Callback URL to call upon completion if the started operation is async.
+	CallbackURL string
+	// The original HTTP request.
+	// Read the URL, Header, and Body of the request to process the operation input.
+	HTTPRequest *http.Request
+}
+
+// GetOperationResultRequest is input for Handler.GetOperationResult.
+type GetOperationResultRequest struct {
+	// Operation name.
+	Operation string
+	// Operation ID as originally generated by a Handler.
+	// It is the handler's responsibility to validate this ID and authorize access to the underlying resource.
+	OperationID string
+	// Whether the caller has indicated that it wants to wait for operation completion, turning the request into a
+	// long poll. When set, the wait duration will be reflected in the provided context deadline.
```

Comment:
Context should be unrelated to wait amount IMO. Users need to differentiate between context close (e.g. server shutdown) and wait complete and to ask them to do that with `ctx.Err()` is a bit much.